src/treetensornetworks/opsum_to_ttn.jl

```diff
@@ -413,10 +428,61 @@ function ttn_svd(
       H[v] += T * ITensorNetworks.computeSiteProd(sites, Prod([(Op("Id", v))]))
     end
   end
-
   return H
 end
 
+#
+# TermQN implements a function with an internal cache.
+# Calling it on a term determines that term's flux
+# but tries not to rebuild the corresponding ITensor.
+# (Previously it was the function calc_qn.)
+#
+struct TermQN{V}
+  sites::IndsNetwork
+  op_cache::Dict{Pair{String,V},ITensor}
+  TermQN{V}(s) where {V} = new{V}(s, Dict{Pair{String,V},ITensor}())
+end
+
+function (t::TermQN)(term)
+  q = QN()
+  for st in term
+    op_tensor = get(t.op_cache, which_op(st) => site(st), nothing)
+    if op_tensor === nothing
+      op_tensor = op(t.sites[site(st)], which_op(st); params(st)...)
+      t.op_cache[which_op(st) => site(st)] = op_tensor
+    end
+    if !isnothing(flux(op_tensor))
+      q += flux(op_tensor)
+    end
+  end
+  return q
+end
+
+function ttn_svd(
+  coefficient_type::Type{<:Number}, os::OpSum, sites::IndsNetwork, root_vertex; kws...
+)
+  calc_qn = TermQN{vertextype(sites)}(sites)
+
+  tempTTN, inbond_coefs = make_sparse_ttn(os, sites, root_vertex, coefficient_type, calc_qn)
+
+  # Traverse tree outwards from root vertex
+  ordered_verts = _default_vertex_ordering(sites, root_vertex)
+  # Store edges in fixed ordering relative to root
+  ordered_edges = _default_edge_ordering(sites, root_vertex)
+
+  Vs = svd_bond_coefs(
+    coefficient_type, sites, ordered_verts, ordered_edges, inbond_coefs; kws...
+  )
+
+  Hflux = -calc_qn(terms(first(terms(os))))
+
+  T = compress_ttn(
+    coefficient_type, sites, ordered_verts, ordered_edges, Hflux, tempTTN, Vs
+  )
```

Comment:
I'm not a fan of this kinds of functions that take many positional arguments, to me it is a warning sign that maybe the number of arguments could be cut down, the functions could be split up into smaller parts or refactored in some other way, or arguments could be moved to keyword arguments.

We could leave that for future work, some of it may be improved naturally by refactoring in terms of better data structures.

Reply:
Totally agree. I was thinking that this 'entanglement' between the different functions could be left for a future PR.